ffmpeg 5 support

### Actual behaviour

NixOS upgraded ffmpeg to 5.0 and the USDX configure script errored out.

```
checking for libavformat... yes (-L/nix/store/jaf390yiaqhwh15na2gw8cdnsngk7g7s-ffmpeg-5.0/lib)
checking version of libavformat... [59.16.100]
checking for libavutil... yes (-L/nix/store/jaf390yiaqhwh15na2gw8cdnsngk7g7s-ffmpeg-5.0/lib)
checking version of libavutil... [57.17.100]
checking for libswscale... yes (-L/nix/store/jaf390yiaqhwh15na2gw8cdnsngk7g7s-ffmpeg-5.0/lib)
checking version of libswscale... [6.4.100]
checking for libswresample... yes (-L/nix/store/jaf390yiaqhwh15na2gw8cdnsngk7g7s-ffmpeg-5.0/lib)
checking version of libswresample... [4.3.100]
checking version of ffmpeg... configure: error:

Unsupported ffmpeg version, most recent version supported is 3.3.
```

It is interesting to note, that we previously used ffmpeg 4, so that error messages seems somewhat out of date.

### Expected behaviour

USDX should be updated to work with ffmpeg 5.0

### Steps to reproduce

Trivial.

### Details

Provide some additional information:

- USDX version: ade03ddd6d995a78c11de0ed3a1f052adc43ff12
- Operating System + version: NixOS 22.05 (unstable)

--- a/configure.ac
+++ b/configure.ac
@@ -300,7 +300,9 @@ AC_SUBST_DEFINE(HAVE_SWRESAMPLE, $libswresample_HAVE)
 # map avutil library version to ffmpeg version
 AC_MSG_CHECKING([version of ffmpeg])
 # the order is ([avutil], [avcodec], [avformat], [swscale], [swresample])
-if   FFMPEG_LIBS_COMPATIBLE([56014100], [58018100], [58012100], [5001100], [3001100]); then
+if   FFMPEG_LIBS_COMPATIBLE([57017100], [59018100], [59016100], [6004100], [4003100]); then
+    FFMPEG_VERSION="5.0"
+elif FFMPEG_LIBS_COMPATIBLE([56014100], [58018100], [58012100], [5001100], [3001100]); then
     FFMPEG_VERSION="4.0"
 elif FFMPEG_LIBS_COMPATIBLE([55078100], [57107100], [57083100], [4008100], [2009100]); then
     FFMPEG_VERSION="3.4"
@@ -347,7 +349,7 @@ elif test $libavutil_VERSION_INT -le 50024000 -a $libavutil_VERSION_INT -ge 4900
 else
     AC_MSG_ERROR([
 
-Unsupported ffmpeg version, most recent version supported is 3.3.
+Unsupported ffmpeg version, most recent version supported is 5.0.
 ])
 fi
 AX_EXTRACT_VERSION(FFMPEG, $FFMPEG_VERSION)

--- a/src/lib/ffmpeg-4.0/avutil.pas
+++ b/src/lib/ffmpeg-4.0/avutil.pas
@@ -95,6 +95,7 @@ type
   cenum = cint;
 
 const
+  AVERROR_EOF = -(ord('E') or (ord('O') shl 8) or (ord('F') shl 16) or (ord(' ') shl 24));
   AV_LOG_FATAL = 8;
   AV_NOPTS_VALUE = -9223372036854775808;
   AV_NUM_DATA_POINTERS = 8;

--- a/src/lib/ffmpeg-5.0/avcodec.pas
+++ b/src/lib/ffmpeg-5.0/avcodec.pas
@@ -1,0 +1,290 @@
+unit avcodec;
+
+{$IFDEF FPC}
+  {$MODE DELPHI}
+  {$PACKENUM 4}    (* use 4-byte enums *)
+  {$PACKRECORDS C} (* C/C++-compatible record packing *)
+{$ELSE}
+  {$MINENUMSIZE 4} (* use 4-byte enums *)
+{$ENDIF}
+
+{$IFDEF DARWIN}
+  {$linklib libavcodec}
+{$ENDIF}
+
+interface
+
+uses
+  ctypes,
+  avutil,
+  rational,
+  SysUtils,
+  UConfig;
+
+const
+  (* Supported version by this header *)
+  LIBAVCODEC_MAX_VERSION_MAJOR   = 59;
+  LIBAVCODEC_MAX_VERSION_MINOR   = 18;
+  LIBAVCODEC_MAX_VERSION_RELEASE = 100;
+  LIBAVCODEC_MAX_VERSION = (LIBAVCODEC_MAX_VERSION_MAJOR * VERSION_MAJOR) +
+                           (LIBAVCODEC_MAX_VERSION_MINOR * VERSION_MINOR) +
+                           (LIBAVCODEC_MAX_VERSION_RELEASE * VERSION_RELEASE);
+
+  (* Min. supported version by this header *)
+  LIBAVCODEC_MIN_VERSION_MAJOR   = 59;
+  LIBAVCODEC_MIN_VERSION_MINOR   = 18;
+  LIBAVCODEC_MIN_VERSION_RELEASE = 100;
+  LIBAVCODEC_MIN_VERSION = (LIBAVCODEC_MIN_VERSION_MAJOR * VERSION_MAJOR) +
+                            (LIBAVCODEC_MIN_VERSION_MINOR * VERSION_MINOR) +
+                            (LIBAVCODEC_MIN_VERSION_RELEASE * VERSION_RELEASE);
+
+(* Check if linked versions are supported *)
+{$IF (LIBAVCODEC_VERSION < LIBAVCODEC_MIN_VERSION)}
+  {$MESSAGE Error 'Linked version of libavcodec is too old!'}
+{$IFEND}
+
+(* Check if linked version is supported *)
+{$IF (LIBAVCODEC_VERSION > LIBAVCODEC_MAX_VERSION)}
+  {$MESSAGE Error 'Linked version of libavcodec is not yet supported!'}
+{$IFEND}
+
+const
+  FF_BUG_AUTODETECT = 1;
+type
+  TAVCodecID = (
+    AV_CODEC_ID_NONE
+  );
+  PAVPacket = ^TAVPacket;
+  TAVPacket = record
+    we_do_not_use_buf: pointer;
+    pts: cint64;
+    we_do_not_use_dts: cint64;
+    data: PByteArray;
+    size: cint;
+    stream_index: cint;
+    flags: cint;
+    we_do_not_use_side_data: pointer;
+    we_do_not_use_side_data_elems: cint;
+    we_do_not_use_duration: cint64;
+    we_do_not_use_pos: cint64;
+    we_do_not_use_opaque: pointer;
+    we_do_not_use_opaque_ref: pointer;
+    we_do_not_use_time_base: TAVRational;
+  end;
+  PAVCodecDescriptor = ^TAVCodecDescriptor;
+  TAVCodecDescriptor = record
+    we_do_not_use_id: TAVCodecID;
+    we_do_not_use_type: TAVMediaType;
+    name: ^AnsiChar;
+    long_name: ^AnsiChar;
+    do_not_instantiate_this_record: incomplete_record;
+  end;
+  PAVCodecParameters = ^TAVCodecParameters;
+  TAVCodecParameters = record
+    codec_type: TAVMediaType;
+    codec_id: TAVCodecID;
+    do_not_instantiate_this_record: incomplete_record;
+  end;
+  PAVCodecContext = ^TAVCodecContext;
+  PPAVCodecContext = ^PAVCodecContext;
+  PAVCodec = ^TAVCodec;
+  TAVCodec = record
+    name: ^AnsiChar;
+    we_do_not_use_long_name: ^AnsiChar;
+    type_: TAVMediaType;
+    id: TAVCodecID;
+    we_do_not_use_capabilities: cint;
+    we_do_not_use_max_lowres: cuint8;
+    we_do_not_use_supported_framerates: ^TAVRational;
+    pix_fmts: ^TAVPixelFormat;
+    we_do_not_use_supported_samplerates: pcint;
+    we_do_not_use_sample_fmts: ^TAVSampleFormat;
+    we_do_not_use_channel_layouts: pcuint64;
+    we_do_not_use_priv_class: pointer;
+    we_do_not_use_profiles: pointer;
+    we_do_not_use_wrapper_name: ^AnsiChar;
+    we_do_not_use_priv_data_size: cint;
+    we_do_not_use_next: PAVCodec;
+{$IF LIBAVCODEC_VERSION < 58079100}
+    we_do_not_use_init_thread_copy: cfunctionpointer;
+{$ENDIF}
+    we_do_not_use_update_thread_context: cfunctionpointer;
+    we_do_not_use_defaults: pointer;
+    we_do_not_use_init_static_data: cfunctionpointer;
+    we_do_not_use_init: cfunctionpointer;
+    we_do_not_use_encode_sub: cfunctionpointer;
+    we_do_not_use_encode2: cfunctionpointer;
+    we_do_not_use_decode: cfunctionpointer;
+    we_do_not_use_close: cfunctionpointer;
+{$IF LIBAVCODEC_VERSION < 58093100}
+    we_do_not_use_send_frame: cfunctionpointer;
+{$ENDIF}
+    we_do_not_use_receive_packet: cfunctionpointer;
+    we_do_not_use_receive_frame: cfunctionpointer;
+    flush: procedure(avctx: PAVCodecContext); cdecl;
+    do_not_instantiate_this_record: incomplete_record;
+  end;
+  TAVCodecContext = record
+    we_do_not_use_av_class: pointer;
+    we_do_not_use_log_level_offset: cint;
+    codec_type: TAVMediaType;
+    codec: ^TAVCodec;
+    codec_id: TAVCodecID;
+    we_do_not_use_codec_tag: cuint;
+    we_do_not_use_priv_data: pointer;
+    we_do_not_use_internal: pointer;
+    we_do_not_use_opaque: pointer;
+    we_do_not_use_bit_rate: cint64;
+    we_do_not_use_bit_rate_tolerance: cint;
+    we_do_not_use_global_quality: cint;
+    we_do_not_use_compression_level: cint;
+    we_do_not_use_flags: cint;
+    we_do_not_use_flags2: cint;
+    we_do_not_use_extradata: pcuint8;
+    we_do_not_use_extradata_size: cint;
+    time_base: TAVRational;
+    we_do_not_use_ticks_per_frame: cint;
+    we_do_not_use_delay: cint;
+    width: cint;
+    height: cint;
+    we_do_not_use_coded_width: cint;
+    we_do_not_use_coded_height: cint;
+    we_do_not_use_gop_size: cint;
+    pix_fmt: TAVPixelFormat;
+    we_do_not_use_draw_horiz_band: cfunctionpointer;
+    get_format: function(s: PAVCodecContext; fmt: PAVPixelFormat): TAVPixelFormat; cdecl;
+    we_do_not_use_max_b_frames: cint;
+    we_do_not_use_b_quant_factor: cfloat;
+    we_do_not_use_b_quant_offset: cfloat;
+    we_do_not_use_has_b_frames: cint;
+    we_do_not_use_i_quant_factor: cfloat;
+    we_do_not_use_i_quant_offset: cfloat;
+    we_do_not_use_lumi_masking: cfloat;
+    we_do_not_use_temporal_cplx_masking: cfloat;
+    we_do_not_use_spatial_cplx_masking: cfloat;
+    we_do_not_use_p_masking: cfloat;
+    we_do_not_use_dark_masking: cfloat;
+    we_do_not_use_slice_count: cint;
+    we_do_not_use_slice_offset: pcint;
+    sample_aspect_ratio: TAVRational;
+    we_do_not_use_me_cmp: cint;
+    we_do_not_use_me_sub_cmp: cint;
+    we_do_not_use_mb_cmp: cint;
+    we_do_not_use_ildct_cmp: cint;
+    we_do_not_use_dia_size: cint;
+    we_do_not_use_last_predictor_count: cint;
+    we_do_not_use_me_pre_cmp: cint;
+    we_do_not_use_pre_dia_size: cint;
+    we_do_not_use_me_subpel_quality: cint;
+    we_do_not_use_me_range: cint;
+    we_do_not_use_slice_flags: cint;
+    we_do_not_use_mb_decision: cint;
+    we_do_not_use_intra_matrix: pcuint16;
+    we_do_not_use_inter_matrix: pcuint16;
+    we_do_not_use_intra_dc_precision: cint;
+    we_do_not_use_skip_top: cint;
+    we_do_not_use_skip_bottom: cint;
+    we_do_not_use_mb_lmin: cint;
+    we_do_not_use_mb_lmax: cint;
+    we_do_not_use_bidir_refine: cint;
+    we_do_not_use_keyint_min: cint;
+    we_do_not_use_refs: cint;
+    we_do_not_use_mv0_threshold: cint;
+    we_do_not_use_color_primaries: cenum;
+    we_do_not_use_color_trc: cenum;
+    we_do_not_use_colorspace: cenum;
+    we_do_not_use_color_range: cenum;
+    we_do_not_use_chroma_sample_location: cenum;
+    we_do_not_use_slices: cint;
+    we_do_not_use_field_order: cenum;
+    sample_rate: cint;
+    channels: cint;
+    sample_fmt: TAVSampleFormat;
+    we_do_not_use_frame_size: cint;
+    we_do_not_use_frame_number: cint;
+    we_do_not_use_block_align: cint;
+    we_do_not_use_cutoff: cint;
+    channel_layout: cuint64;
+    request_channel_layout: cuint64;
+    we_do_not_use_audio_service_type: cenum;
+    request_sample_fmt: TAVSampleFormat;
+    we_do_not_use_get_buffer2: cfunctionpointer;
+    we_do_not_use_qcompress: cfloat;
+    we_do_not_use_qblur: cfloat;
+    we_do_not_use_qmin: cint;
+    we_do_not_use_qmax: cint;
+    we_do_not_use_max_qdiff: cint;
+    we_do_not_use_rc_buffer_size: cint;
+    we_do_not_use_rc_override_count: cint;
+    we_do_not_use_rc_override: pointer;
+    we_do_not_use_rc_max_rate: cint64;
+    we_do_not_use_rc_min_rate: cint64;
+    we_do_not_use_rc_max_available_vbv_use: cfloat;
+    we_do_not_use_rc_min_vbv_overflow_use: cfloat;
+    we_do_not_use_rc_initial_buffer_occupancy: cint;
+    we_do_not_use_trellis: cint;
+    we_do_not_use_stats_out: pcchar;
+    we_do_not_use_stats_in: pcchar;
+    workaround_bugs: cint;
+    we_do_not_use_strict_std_compliance: cint;
+    we_do_not_use_error_concealment: cint;
+    debug: cint;
+    we_do_not_use_err_recognition: cint;
+    we_do_not_use_reordered_opaque: cint64;
+    we_do_not_use_hwaccel: pointer;
+    we_do_not_use_hwaccel_context: pointer;
+    we_do_not_use_error: array [0..AV_NUM_DATA_POINTERS-1] of cuint64;
+    we_do_not_use_dct_algo: cint;
+    we_do_not_use_idct_algo: cint;
+    we_do_not_use_bits_per_coded_sample: cint;
+    we_do_not_use_bits_per_raw_sample: cint;
+    we_do_not_use_lowres: cint;
+    we_do_not_use_thread_count: cint;
+    we_do_not_use_thread_type: cint;
+    we_do_not_use_active_thread_type: cint;
+    we_do_not_use_thread_safe_callbacks: cint;
+    we_do_not_use_execute: cfunctionpointer;
+    we_do_not_use_execute2: cfunctionpointer;
+    we_do_not_use_nsse_weight: cint;
+    we_do_not_use_profile: cint;
+    we_do_not_use_level: cint;
+    we_do_not_use_skip_loop_filter: cenum;
+    we_do_not_use_skip_idct: cenum;
+    we_do_not_use_skip_frame: cenum;
+    we_do_not_use_subtitle_header: pcuint8;
+    we_do_not_use_subtitle_header_size: cint;
+    we_do_not_use_initial_padding: cint;
+    framerate: TAVRational;
+    we_do_not_use_sw_pix_fmt: TAVPixelFormat;
+    we_do_not_use_pkt_timebase: TAVRational;
+    we_do_not_use_codec_descriptor: PAVCodecDescriptor;
+    we_do_not_use_pts_correction_num_faulty_pts: cint64;
+    we_do_not_use_pts_correction_num_faulty_dts: cint64;
+    we_do_not_use_pts_correction_last_pts: cint64;
+    we_do_not_use_pts_correction_last_dts: cint64;
+    we_do_not_use_sub_charenc: pcchar;
+    we_do_not_use_sub_charenc_mode: cint;
+    we_do_not_use_skip_alpha: cint;
+    we_do_not_use_seek_preroll: cint;
+    debug_mv: cint;
+    do_not_instantiate_this_record: incomplete_record;
+  end;
+function av_packet_ref(dst: PAVPacket; src: PAVPacket): cint; cdecl; external av__codec;
+procedure av_packet_unref(pkt: PAVPacket); cdecl; external av__codec;
+procedure av_init_packet(var pkt: TAVPacket); cdecl; external av__codec; deprecated;
+function avcodec_version(): cuint; cdecl; external av__codec;
+function av_codec_is_decoder(codec: PAVCodec): cint; cdecl; external av__codec;
+function av_codec_iterate(opaque: ppointer): PAVCodec; cdecl; external av__codec;
+function avcodec_find_decoder(id: TAVCodecID): PAVCodec; cdecl; external av__codec;
+function avcodec_find_decoder_by_name(name: PAnsiChar): PAVCodec; cdecl; external av__codec;
+function avcodec_descriptor_get(id: TAVCodecID): PAVCodecDescriptor; cdecl; external av__codec;
+function avcodec_open2(avctx: PAVCodecContext; codec: PAVCodec; options: PPAVDictionary): cint; cdecl; external av__codec;
+function avcodec_close(avctx: PAVCodecContext): cint; cdecl; external av__codec;
+procedure avcodec_flush_buffers(avctx: PAVCodecContext); cdecl; external av__codec;
+function avcodec_receive_frame(avctx: PAVCodecContext; frame: PAVFrame): cint; cdecl; external av__codec;
+function avcodec_send_packet(avctx: PAVCodecContext; avpkt: PAVPacket): cint; cdecl; external av__codec;
+function avcodec_alloc_context3(codec: PAVCodec): PAVCodecContext; cdecl; external av__codec;
+procedure avcodec_free_context(avctx: PPAVCodecContext); cdecl; external av__codec;
+function avcodec_parameters_to_context(codec: PAVCodecContext; par: PAVCodecParameters): cint; cdecl; external av__codec;
+implementation
+end.

--- a/src/lib/ffmpeg-5.0/avformat.pas
+++ b/src/lib/ffmpeg-5.0/avformat.pas
@@ -1,0 +1,118 @@
+unit avformat;
+
+{$IFDEF FPC}
+  {$MODE DELPHI}
+  {$PACKENUM 4}    (* use 4-byte enums *)
+  {$PACKRECORDS C} (* C/C++-compatible record packing *)
+{$ELSE}
+  {$MINENUMSIZE 4} (* use 4-byte enums *)
+{$ENDIF}
+
+{$IFDEF DARWIN}
+  {$linklib libavformat}
+{$ENDIF}
+
+interface
+
+uses
+  ctypes,
+  avcodec,
+  avio,
+  avutil,
+  rational,
+  UConfig;
+
+const
+  (* Max. supported version by this header *)
+  LIBAVFORMAT_MAX_VERSION_MAJOR   = 59;
+  LIBAVFORMAT_MAX_VERSION_MINOR   = 16;
+  LIBAVFORMAT_MAX_VERSION_RELEASE = 100;
+  LIBAVFORMAT_MAX_VERSION = (LIBAVFORMAT_MAX_VERSION_MAJOR * VERSION_MAJOR) +
+                            (LIBAVFORMAT_MAX_VERSION_MINOR * VERSION_MINOR) +
+                            (LIBAVFORMAT_MAX_VERSION_RELEASE * VERSION_RELEASE);
+
+  (* Min. supported version by this header *)
+  LIBAVFORMAT_MIN_VERSION_MAJOR   = 59;
+  LIBAVFORMAT_MIN_VERSION_MINOR   = 16;
+  LIBAVFORMAT_MIN_VERSION_RELEASE = 100;
+  LIBAVFORMAT_MIN_VERSION = (LIBAVFORMAT_MIN_VERSION_MAJOR * VERSION_MAJOR) +
+                            (LIBAVFORMAT_MIN_VERSION_MINOR * VERSION_MINOR) +
+                            (LIBAVFORMAT_MIN_VERSION_RELEASE * VERSION_RELEASE);
+
+(* Check if linked versions are supported *)
+{$IF (LIBAVFORMAT_VERSION < LIBAVFORMAT_MIN_VERSION)}
+  {$MESSAGE Error 'Linked version of libavformat is too old!'}
+{$IFEND}
+
+(* Check if linked versions are supported *)
+{$IF (LIBAVFORMAT_VERSION > LIBAVFORMAT_MAX_VERSION)}
+  {$MESSAGE Error 'Linked version of libavformat is not yet supported!'}
+{$IFEND}
+
+const
+  AVFMT_FLAG_GENPTS = 1;
+  AVSEEK_FLAG_ANY = 4;
+  AVSEEK_FLAG_BACKWARD = 1;
+type
+  PAVPacketList = ^TAVPacketList;
+  TAVPacketList = record
+    pkt: TAVPacket;
+    next: ^TAVPacketList;
+  end;
+  PAVInputFormat = ^TAVInputFormat;
+  PAVStream = ^TAVStream;
+  PPAVStream = ^PAVStream;
+  PAVFormatContext = ^TAVFormatContext;
+  PPAVFormatContext = ^PAVFormatContext;
+  TAVFormatContext = record
+    we_do_not_use_av_class: pointer;
+    iformat: PAVInputFormat;
+    we_do_not_use_oformat: pointer;
+    we_do_not_use_priv_data: pointer;
+    pb: ^TAVIOContext;
+    we_do_not_use_ctx_flags: cint;
+    nb_streams: cuint;
+    streams: PPAVStream;
+    url: ^AnsiChar;
+    start_time: cint64;
+    duration: cint64;
+    we_do_not_use_bit_rate: cint64;
+    we_do_not_use_packet_size: cuint;
+    we_do_not_use_max_delay: cint;
+    flags: cint;
+    do_not_instantiate_this_record: incomplete_record;
+  end;
+  TAVStream = record
+    we_do_not_use_index: cint;
+    we_do_not_use_id: cint;
+    we_do_not_use_priv_data: pointer;
+    time_base: TAVRational;
+    start_time: cint64;
+    we_do_not_use_duration: cint64;
+    we_do_not_use_nb_frames: cint64;
+    we_do_not_use_disposition: cint;
+    we_do_not_use_discard: cenum;
+    we_do_not_use_sample_aspect_ratio: TAVRational;
+    we_do_not_use_metadata: PAVDictionary;
+    we_do_not_use_avg_frame_rate: TAVRational;
+    we_do_not_use_atached_pic: TAVPacket;
+    we_do_not_use_side_data: pointer;
+    we_do_not_use_nb_side_data: cint;
+    we_do_not_use_event_flags: cint;
+    r_frame_rate: TAVRational;
+    codecpar: ^TAVCodecParameters;
+    do_not_instantiate_this_record: incomplete_record;
+  end;
+  TAVInputFormat = record
+    name: ^AnsiChar;
+    do_not_instantiate_this_record: incomplete_record;
+  end;
+function avformat_alloc_context(): PAVFormatContext; cdecl; external av__format;
+function avformat_open_input(ps: PPAVFormatContext; url: PAnsiChar; fmt: PAVInputFormat; options: PPAVDictionary): cint; cdecl; external av__format;
+procedure avformat_close_input(s: PPAVFormatContext); cdecl; external av__format;
+function avformat_version(): cuint; cdecl; external av__format;
+function avformat_find_stream_info(ic: PAVFormatContext; options: PPAVDictionary): cint; cdecl; external av__format;
+function av_read_frame(s: PAVFormatContext; var pkt: TAVPacket): cint; cdecl; external av__format;
+function av_seek_frame(s: PAVFormatContext; stream_index: cint; timestamp: cint64; flags: cint): cint; cdecl; external av__format;
+implementation
+end.

--- a/src/lib/ffmpeg-5.0/avio.pas
+++ b/src/lib/ffmpeg-5.0/avio.pas
@@ -1,0 +1,47 @@
+unit avio;
+
+{$IFDEF FPC}
+  {$MODE DELPHI}
+  {$PACKENUM 4}    (* use 4-byte enums *)
+  {$PACKRECORDS C} (* C/C++-compatible record packing *)
+{$ELSE}
+  {$MINENUMSIZE 4} (* use 4-byte enums *)
+{$ENDIF}
+
+{$IFDEF DARWIN}
+  {$linklib libavformat}
+{$ENDIF}
+
+interface
+
+uses
+  ctypes,
+  avutil,
+  SysUtils,
+  UConfig;
+const
+  AVSEEK_SIZE = 65536;
+type
+  PAVIOContext = ^TAVIOContext;
+  TAVIOContext = record
+    we_do_not_use_av_class: pointer;
+    buffer: ^cuchar;
+    we_do_not_use_buffer_size: cint;
+    we_do_not_use_buf_ptr: pcuchar;
+    we_do_not_use_buf_end: pcuchar;
+    opaque: pointer;
+    we_do_not_use_read_packet: cfunctionpointer;
+    we_do_not_use_write_packet: cfunctionpointer;
+    we_do_not_use_seek: cfunctionpointer;
+    pos: cint64;
+    eof_reached: cint;
+    error: cint;
+    do_not_instantiate_this_record: incomplete_record;
+  end;
+  TReadWriteFunc = function(opaque: pointer; buf: pbytearray; buf_size: cint): cint; cdecl;
+  TSeekFunc = function(opaque: pointer; offset: cint64; whence: cint): cint64; cdecl;
+function avio_alloc_context(buffer: pcuint8; buffer_size: cint; write_flag: cint; opaque: pointer; read_packet: TReadWriteFunc; write_packet: TReadWriteFunc; seek: TSeekFunc): PAVIOContext; cdecl; external av__format;
+function avio_feof(s: PAVIOContext): cint; cdecl; external av__format;
+function avio_size(s: PAVIOContext): cint64; cdecl; external av__format;
+implementation
+end.

--- a/src/lib/ffmpeg-5.0/avutil.pas
+++ b/src/lib/ffmpeg-5.0/avutil.pas
@@ -1,0 +1,186 @@
+unit avutil;
+
+{$IFDEF FPC}
+  {$MODE DELPHI}
+  {$MODESWITCH ADVANCEDRECORDS}
+  {$PACKENUM 4}    (* use 4-byte enums *)
+  {$PACKRECORDS C} (* C/C++-compatible record packing *)
+{$ELSE}
+  {$MINENUMSIZE 4} (* use 4-byte enums *)
+{$ENDIF}
+
+{$IFDEF DARWIN}
+  {$linklib libavutil}
+{$ENDIF}
+
+interface
+
+uses
+  ctypes,
+  rational,
+  {$IFDEF UNIX}
+  BaseUnix,
+  {$ENDIF}
+  SysUtils,
+  UConfig;
+
+
+const
+  (* Max. supported version by this header *)
+  LIBAVUTIL_MAX_VERSION_MAJOR   = 57;
+  LIBAVUTIL_MAX_VERSION_MINOR   = 17;
+  LIBAVUTIL_MAX_VERSION_RELEASE = 100;
+  LIBAVUTIL_MAX_VERSION = (LIBAVUTIL_MAX_VERSION_MAJOR * VERSION_MAJOR) +
+                          (LIBAVUTIL_MAX_VERSION_MINOR * VERSION_MINOR) +
+                          (LIBAVUTIL_MAX_VERSION_RELEASE * VERSION_RELEASE);
+
+  (* Min. supported version by this header *)
+  LIBAVUTIL_MIN_VERSION_MAJOR   = 57;
+  LIBAVUTIL_MIN_VERSION_MINOR   = 17;
+  LIBAVUTIL_MIN_VERSION_RELEASE = 100;
+  LIBAVUTIL_MIN_VERSION = (LIBAVUTIL_MIN_VERSION_MAJOR * VERSION_MAJOR) +
+                          (LIBAVUTIL_MIN_VERSION_MINOR * VERSION_MINOR) +
+                          (LIBAVUTIL_MIN_VERSION_RELEASE * VERSION_RELEASE);
+
+(* Check if linked versions are supported *)
+{$IF (LIBAVUTIL_VERSION < LIBAVUTIL_MIN_VERSION)}
+  {$MESSAGE Error 'Linked version of libavutil is too old!'}
+{$IFEND}
+
+{$IF (LIBAVUTIL_VERSION > LIBAVUTIL_MAX_VERSION)}
+  {$MESSAGE Error 'Linked version of libavutil is not yet supported!'}
+{$IFEND}
+
+type
+  (*
+   * We use this record as the last element of records that are not supposed
+   * to be allocated by pascal code either because the declaration of the
+   * record has never been completed or because FFmpeg does the allocation
+   * internally.
+   *
+   * The exceed_stack array has been declared as 10MB array because the default
+   * soft stack limit on Linux is less than that. So every time someone tries
+   * to allocate this structure on the stack, it will crash unless the soft
+   * limit has been raised.
+   *
+   * The existence of a constructor will cause Free Pascal to issue a warning
+   * if an element of this type has been created without calling the
+   * constructor.
+   *)
+  incomplete_record = record
+    exceed_stack: array [0..10000000] of byte;
+    constructor Never_Call(dummy: integer);
+  end;
+
+type
+  (*
+   * The C standard allows for function pointers to have a different size than
+   * data pointers. In practice the size differs only on exotic architectures.
+   *)
+  cfunctionpointer = procedure;cdecl;
+  (*
+   * Here we assume that the size of an enum is the same regardless of the
+   * values it needs to hold. This is not always the case. The AMD64 System V
+   * ABI says:
+   *
+   *     C++ and some implementations of C permit enums larger than an int.
+   *     The underlying type is bumped to an unsigned int, long int or
+   *     unsigned long int, in that order.
+   *
+   * Apart from that there are systems where the default enum size is smaller
+   * than an int. But these are unlikely to run FFmpeg.
+   *
+   * So don't use this type to replace enums that don't fit into 32 bits.
+   *)
+  cenum = cint;
+
+const
+  AVERROR_EOF = -(ord('E') or (ord('O') shl 8) or (ord('F') shl 16) or (ord(' ') shl 24));
+  AV_LOG_FATAL = 8;
+  AV_NOPTS_VALUE = -9223372036854775808;
+  AV_NUM_DATA_POINTERS = 8;
+  AV_TIME_BASE = 1000000;
+{$IFDEF UNIX}
+  EAGAIN = ESysEAGAIN;
+{$ELSE}
+  EAGAIN = 11;
+{$ENDIF}
+
+type
+  TAVSampleFormat = (
+    AV_SAMPLE_FMT_U8,
+    AV_SAMPLE_FMT_S16,
+    AV_SAMPLE_FMT_S32,
+    AV_SAMPLE_FMT_FLT,
+    AV_SAMPLE_FMT_DBL
+  );
+  PAVPixelFormat = ^TAVPixelFormat;
+  TAVPixelFormat = (
+    AV_PIX_FMT_RGB24 = 2,
+    AV_PIX_FMT_BGR24,
+    AV_PIX_FMT_RGBA = 26,
+    AV_PIX_FMT_BGRA = 28
+  );
+  TAVMediaType = (
+    AVMEDIA_TYPE_VIDEO,
+    AVMEDIA_TYPE_AUDIO
+  );
+  ppcuint8 = ^pcuint8;
+  PAVFrame = ^TAVFrame;
+  PPAVFrame = ^PAVFrame;
+  TAVFrame = record
+    data: array [0..AV_NUM_DATA_POINTERS-1] of ^cuint8;
+    linesize: array [0..AV_NUM_DATA_POINTERS-1] of cint;
+    extended_data: ^pcuint8;
+    we_do_not_use_width: cint;
+    we_do_not_use_height: cint;
+    nb_samples: cint;
+    we_do_not_use_format: cint;
+    we_do_not_use_key_frame: cint;
+    we_do_not_use_pict_type: cenum;
+    we_do_not_use_sample_aspect_ratio: TAVRational;
+    pts: cint64;
+    pkt_dts: cint64;
+    we_do_not_use_time_base: TAVRational;
+    we_do_not_use_coded_picture_number: cint;
+    we_do_not_use_display_picture_number: cint;
+    we_do_not_use_quality: cint;
+    we_do_not_use_opaque: pointer;
+    repeat_pict: cint;
+    do_not_instantiate_this_record: incomplete_record;
+  end;
+  PAVDictionary = ^TAVDictionary;
+  PPAVDictionary = ^PAVDictionary;
+  TAVDictionary = record
+    do_not_instantiate_this_record: incomplete_record;
+  end;
+procedure av_free(ptr: pointer); cdecl; external av__util;
+procedure av_freep(ptr: pointer); cdecl; external av__util;
+function av_malloc(size: csize_t): pointer; cdecl; external av__util;
+function avutil_version(): cuint; cdecl; external av__util;
+function av_frame_alloc(): PAVFrame; cdecl; external av__util;
+procedure av_frame_free(frame: PPAVFrame); cdecl; external av__util;
+function av_image_alloc(pointers: ppcuint8; linesizes: pcint; w: cint; h: cint; pix_fmt: TAVPixelFormat; align: cint): cint; cdecl; external av__util;
+function AVERROR(e: cint): cint; {$IFDEF HasInline}inline;{$ENDIF}
+function av_opt_set_int(obj: pointer; name: PAnsiChar; val: cint64; search_flags: cint): cint; cdecl; external av__util;
+function av_opt_set_channel_layout(obj: pointer; name: PAnsiChar; ch_layout: cint64; search_flags: cint): cint; cdecl; external av__util;
+function av_opt_set_sample_fmt(obj: pointer; name: PAnsiChar; fmt: TAVSampleFormat; search_flags: cint): cint; cdecl; external av__util;
+function av_samples_alloc(var audio_data: pcuint8; linesize: pcint; nb_channels: cint; nb_samples: cint; sample_fmt: TAVSampleFormat; align: cint): cint; cdecl; external av__util;
+function av_get_packed_sample_fmt(sample_fmt: TAVSampleFormat): TAVSampleFormat; cdecl; external av__util;
+function av_get_bytes_per_sample(sample_fmt: TAVSampleFormat): cint; cdecl; external av__util;
+procedure av_log_set_level(level: cint); cdecl; external av__util;
+implementation
+function AVERROR(e: cint): cint; {$IFDEF HasInline}inline;{$ENDIF}
+begin
+  if EAGAIN < 0 then
+    AVERROR := e
+  else
+    AVERROR := -e;
+end;
+
+constructor incomplete_record.Never_Call(dummy: integer);
+begin
+  abort;
+end;
+
+end.

--- a/src/lib/ffmpeg-5.0/rational.pas
+++ b/src/lib/ffmpeg-5.0/rational.pas
@@ -1,0 +1,32 @@
+unit rational;
+
+{$IFDEF FPC}
+  {$MODE DELPHI}
+  {$PACKENUM 4}    (* use 4-byte enums *)
+  {$PACKRECORDS C} (* C/C++-compatible record packing *)
+{$ELSE}
+  {$MINENUMSIZE 4} (* use 4-byte enums *)
+{$ENDIF}
+
+interface
+
+uses
+  ctypes;
+type
+  TAVRational = record
+    num: csint;
+    den: csint;
+  end;
+function av_q2d(a: TAVRational): cdouble; {$IFDEF HasInline}inline;{$ENDIF}
+function av_inv_q(q: TAVRational): TAVRational; {$IFDEF HasInline}inline;{$ENDIF}
+implementation
+function av_q2d(a: TAVRational): cdouble; {$IFDEF HasInline}inline;{$ENDIF}
+begin
+  av_q2d := a.num / a.den;
+end;
+function av_inv_q(q: TAVRational): TAVRational; {$IFDEF HasInline}inline;{$ENDIF}
+begin
+  av_inv_q.num := q.den;
+  av_inv_q.den := q.num;
+end;
+end.

--- a/src/lib/ffmpeg-5.0/swresample.pas
+++ b/src/lib/ffmpeg-5.0/swresample.pas
@@ -1,0 +1,61 @@
+unit swresample;
+
+{$IFDEF FPC}
+  {$MODE DELPHI}
+  {$PACKENUM 4}    (* use 4-byte enums *)
+  {$PACKRECORDS C} (* C/C++-compatible record packing *)
+{$ELSE}
+  {$MINENUMSIZE 4} (* use 4-byte enums *)
+{$ENDIF}
+
+{$IFDEF DARWIN}
+  {$linklib libswresample}
+{$ENDIF}
+
+interface
+
+uses
+  ctypes,
+  avutil,
+  UConfig;
+
+const
+  (* Supported version by this header *)
+  LIBSWRESAMPLE_MAX_VERSION_MAJOR   = 4;
+  LIBSWRESAMPLE_MAX_VERSION_MINOR   = 3;
+  LIBSWRESAMPLE_MAX_VERSION_RELEASE = 100;
+  LIBSWRESAMPLE_MAX_VERSION = (LIBSWRESAMPLE_MAX_VERSION_MAJOR * VERSION_MAJOR) +
+                           (LIBSWRESAMPLE_MAX_VERSION_MINOR * VERSION_MINOR) +
+                           (LIBSWRESAMPLE_MAX_VERSION_RELEASE * VERSION_RELEASE);
+
+  (* Min. supported version by this header *)
+  LIBSWRESAMPLE_MIN_VERSION_MAJOR   = 0;
+  LIBSWRESAMPLE_MIN_VERSION_MINOR   = 15;
+  LIBSWRESAMPLE_MIN_VERSION_RELEASE = 100;
+  LIBSWRESAMPLE_MIN_VERSION = (LIBSWRESAMPLE_MIN_VERSION_MAJOR * VERSION_MAJOR) +
+                            (LIBSWRESAMPLE_MIN_VERSION_MINOR * VERSION_MINOR) +
+                            (LIBSWRESAMPLE_MIN_VERSION_RELEASE * VERSION_RELEASE);
+
+(* Check if linked versions are supported *)
+{$IF (LIBSWRESAMPLE_VERSION < LIBSWRESAMPLE_MIN_VERSION)}
+  {$MESSAGE Error 'Linked version of libswresample is too old!'}
+{$IFEND}
+
+(* Check if linked version is supported *)
+{$IF (LIBSWRESAMPLE_VERSION > LIBSWRESAMPLE_MAX_VERSION)}
+  {$MESSAGE Error 'Linked version of libswresample is not yet supported!'}
+{$IFEND}
+
+type
+  PSwrContext = ^TSwrContext;
+  PPSwrContext = ^PSwrContext;
+  TSwrContext = record
+    do_not_instantiate_this_record: incomplete_record;
+  end;
+function swr_alloc(): PSwrContext; cdecl; external sw__resample;
+function swr_alloc_set_opts(s: PSwrContext; out_ch_layout: cint64; out_sample_fmt: TAVSampleFormat; out_sample_rate: cint; in_ch_layout: cint64; in_sample_fmt: TAVSampleFormat; in_sample_rate: cint; log_offset: cint; log_ctx: pointer): PSwrContext; cdecl; external sw__resample;
+function swr_init(s: PSwrContext): cint; cdecl; external sw__resample;
+procedure swr_free(s: PPSwrContext); cdecl; external sw__resample;
+function swr_convert(s: PSwrContext; var out: pcuint8; out_count: cint; var in_: pcuint8; in_count: cint): cint; cdecl; external sw__resample;
+implementation
+end.

--- a/src/lib/ffmpeg-5.0/swscale.pas
+++ b/src/lib/ffmpeg-5.0/swscale.pas
@@ -1,0 +1,53 @@
+unit swscale;
+
+{$IFDEF FPC}
+  {$MODE DELPHI}
+  {$PACKENUM 4}    (* use 4-byte enums *)
+  {$PACKRECORDS C} (* C/C++-compatible record packing *)
+{$ELSE}
+  {$MINENUMSIZE 4} (* use 4-byte enums *)
+{$ENDIF}
+
+{$IFDEF DARWIN}
+  {$linklib libswscale}
+{$ENDIF}
+
+interface
+
+uses
+  ctypes,
+  avutil,
+  UConfig;
+
+const
+  (* Max. supported version by this header *)
+  LIBSWSCALE_MAX_VERSION_MAJOR   =  6;
+  LIBSWSCALE_MAX_VERSION_MINOR   =  4;
+  LIBSWSCALE_MAX_VERSION_RELEASE =  100;
+  LIBSWSCALE_MAX_VERSION = (LIBSWSCALE_MAX_VERSION_MAJOR * VERSION_MAJOR) +
+                           (LIBSWSCALE_MAX_VERSION_MINOR * VERSION_MINOR) +
+                           (LIBSWSCALE_MAX_VERSION_RELEASE * VERSION_RELEASE);
+
+(* Check if linked versions are supported *)
+{$IF (LIBSWSCALE_VERSION > LIBSWSCALE_MAX_VERSION)}
+  {$MESSAGE Error 'Linked version of libswscale is not yet supported!'}
+{$IFEND}
+
+
+const
+  SWS_FAST_BILINEAR = 1;
+type
+  PSwsContext = ^TSwsContext;
+  TSwsContext = record
+    do_not_instantiate_this_record: incomplete_record;
+  end;
+  PSwsFilter = ^TSwsFilter;
+  TSwsFilter = record
+    do_not_instantiate_this_record: incomplete_record;
+  end;
+function swscale_version(): cuint; cdecl; external sw__scale;
+function sws_isSupportedInput(pix_fmt: TAVPixelFormat): cint; cdecl; external sw__scale;
+function sws_scale(c: PSwsContext; srcSlice: ppcuint8; srcStride: pcint; srcSliceY: cint; srcSliceH: cint; dst: ppcuint8; dstStride: pcint): cint; cdecl; external sw__scale;
+function sws_getContext(srcW: cint; srcH: cint; srcFormat: TAVPixelFormat; dstW: cint; dstH: cint; dstFormat: TAVPixelFormat; flags: cint; srcFilter: PSwsFilter; dstFilter: PSwsFilter; param: pcdouble): PSwsContext; cdecl; external sw__scale;
+implementation
+end.

--- a/src/media/UAudioDecoder_FFmpeg.pas
+++ b/src/media/UAudioDecoder_FFmpeg.pas
@@ -860,7 +860,11 @@ begin
           begin
             // seeking failed
             fErrorState := true;
+            {$IF LIBAVFORMAT_VERSION <= 58007100}
             Log.LogError('Seek Error in "'+fFormatCtx^.filename+'"', 'UAudioDecoder_FFmpeg');
+            {$ELSE}
+            Log.LogError('Seek Error in "'+fFormatCtx^.url+'"', 'UAudioDecoder_FFmpeg');
+            {$ENDIF}
           end
           else
           begin

--- a/src/media/UAudioDecoder_FFmpeg.pas
+++ b/src/media/UAudioDecoder_FFmpeg.pas
@@ -287,6 +287,10 @@ begin
   fDecoderPauseRequestCount := 0;
 
   FillChar(fAudioPaket, SizeOf(TAVPacket), 0);
+  {$IF (LIBAVFORMAT_VERSION >= 59000000)}
+  // avoid calling av_packet_unref before fetching first frame
+  fAudioPaket.data := Pointer(STATUS_PACKET);
+  {$ENDIF}
 end;
 
 {*
@@ -973,7 +977,11 @@ begin
       if (Packet.stream_index = fAudioStreamIndex) then
         fPacketQueue.Put(@Packet)
       else
+        {$IF (LIBAVFORMAT_VERSION < 59000000)}
         av_free_packet(@Packet);
+        {$ELSE}
+        av_packet_unref(@Packet);
+        {$ENDIF}
 
     finally
       SDL_LockMutex(fStateLock);
@@ -1151,8 +1159,13 @@ begin
     end;
 
     // free old packet data
+    {$IF (LIBAVFORMAT_VERSION < 59000000)}
     if (fAudioPaket.data <> nil) then
       av_free_packet(@fAudioPaket);
+    {$ELSE}
+    if (PAnsiChar(fAudioPaket.data) <> STATUS_PACKET) then
+      av_packet_unref(@fAudioPaket);
+    {$ENDIF}
 
     // do not block queue on seeking (to avoid deadlocks on the DecoderLock)
     if (IsSeeking()) then
@@ -1168,7 +1181,9 @@ begin
     // handle Status-packet
     if (PAnsiChar(fAudioPaket.data) = STATUS_PACKET) then
     begin
+      {$IF (LIBAVFORMAT_VERSION < 59000000)}
       fAudioPaket.data := nil;
+      {$ENDIF}
       fAudioPaketData := nil;
       fAudioPaketSize := 0;
 

--- a/src/media/UAudioDecoder_FFmpeg.pas
+++ b/src/media/UAudioDecoder_FFmpeg.pas
@@ -1356,7 +1356,9 @@ function TAudioDecoder_FFmpeg.InitializeDecoder: boolean;
 begin
   //Log.LogStatus('InitializeDecoder', 'UAudioDecoder_FFmpeg');
   FFmpegCore := TMediaCore_FFmpeg.GetInstance();
+  {$IF LIBAVFORMAT_VERSION < 58027100}
   av_register_all();
+  {$ENDIF}
 
   // Do not show uninformative error messages by default.
   // FFmpeg prints all error-infos on the console by default what

--- a/src/media/UAudioDecoder_FFmpeg.pas
+++ b/src/media/UAudioDecoder_FFmpeg.pas
@@ -407,7 +407,21 @@ begin
   CodecID := fAudioStream^.codecpar^.codec_id;
 {$ENDIF}
 
-  fCodecCtx := fAudioStream^.codec;
+  fCodec := avcodec_find_decoder(CodecID);
+  if (fCodec = nil) then
+  begin
+    Log.LogError('Unsupported codec!', 'UAudio_FFmpeg');
+    fCodecCtx := nil;
+    Close();
+    Exit;
+  end;
+
+  fCodecCtx := FFmpegCore.GetCodecContext(fAudioStream, fCodec);
+  if fCodecCtx = nil then
+  begin
+    Close();
+    Exit;
+  end;
 
   // TODO: should we use this or not? Should we allow 5.1 channel audio?
 
@@ -419,15 +433,6 @@ begin
   else
     fCodecCtx^.request_channels := 2;
   {$IFEND}
-
-  fCodec := avcodec_find_decoder(CodecID);
-  if (fCodec = nil) then
-  begin
-    Log.LogError('Unsupported codec!', 'UAudio_FFmpeg');
-    fCodecCtx := nil;
-    Close();
-    Exit;
-  end;
 
   // set debug options
   fCodecCtx^.debug_mv := 0;
@@ -559,11 +564,17 @@ begin
     // avcodec_close() is not thread-safe
     FFmpegCore.LockAVCodec();
     try
+      {$IF LIBAVFORMAT_VERSION < 59000000}
       avcodec_close(fCodecCtx);
+      {$ELSE}
+      avcodec_free_context(@fCodecCtx);
+      {$ENDIF}
     finally
       FFmpegCore.UnlockAVCodec();
     end;
+    {$IF LIBAVFORMAT_VERSION < 59000000}
     fCodecCtx := nil;
+    {$ENDIF}
   end;
 
   // Close the video file
@@ -1039,6 +1050,10 @@ begin
 end;
 
 procedure TFFmpegDecodeStream.FlushCodecBuffers();
+{$IF LIBAVFORMAT_VERSION >= 59000000}
+var
+  NewCtx: PAVCodecContext;
+{$ENDIF}
 begin
   // if no flush operation is specified, avcodec_flush_buffers will not do anything.
   if (@fCodecCtx.codec.flush <> nil) then
@@ -1053,12 +1068,24 @@ begin
     // We will just reopen the codec.
     FFmpegCore.LockAVCodec();
     try
+      {$IF LIBAVFORMAT_VERSION < 59000000}
       avcodec_close(fCodecCtx);
       {$IF LIBAVCODEC_VERSION >= 53005000}
       avcodec_open2(fCodecCtx, fCodec, nil);
       {$ELSE}
       avcodec_open(fCodecCtx, fCodec);
       {$IFEND}
+      {$ELSE}
+      NewCtx := FFmpegCore.GetCodecContext(fAudioStream, fCodec);
+      if NewCtx <> nil then
+      begin
+        avcodec_free_context(@fCodecCtx);
+        fCodecCtx := NewCtx;
+        avcodec_open2(fCodecCtx, fCodec, nil);
+      end
+      else
+        avcodec_flush_buffers(fCodecCtx);
+      {$ENDIF}
     finally
       FFmpegCore.UnlockAVCodec();
     end;

--- a/src/media/UAudioDecoder_FFmpeg.pas
+++ b/src/media/UAudioDecoder_FFmpeg.pas
@@ -328,6 +328,7 @@ var
   PackedSampleFormat: TAVSampleFormat;
   TestFrame: TAVPacket;
   AVResult: integer;
+  CodecID: TAVCodecID;
 begin
   Result := false;
 
@@ -399,6 +400,13 @@ begin
   fAudioStream := PPAVStream(PtrUInt(fFormatCtx.streams) + fAudioStreamIndex * Sizeof(pointer))^;
 {$IFEND}
   fAudioStreamPos := 0;
+
+{$IF LIBAVFORMAT_VERSION < 59000000}
+  CodecID := fAudioStream^.codec^.codec_id;
+{$ELSE}
+  CodecID := fAudioStream^.codecpar^.codec_id;
+{$ENDIF}
+
   fCodecCtx := fAudioStream^.codec;
 
   // TODO: should we use this or not? Should we allow 5.1 channel audio?
@@ -412,7 +420,7 @@ begin
     fCodecCtx^.request_channels := 2;
   {$IFEND}
 
-  fCodec := avcodec_find_decoder(fCodecCtx^.codec_id);
+  fCodec := avcodec_find_decoder(CodecID);
   if (fCodec = nil) then
   begin
     Log.LogError('Unsupported codec!', 'UAudio_FFmpeg');

--- a/src/media/UMediaCore_FFmpeg.pas
+++ b/src/media/UMediaCore_FFmpeg.pas
@@ -464,6 +464,8 @@ begin
   except
     Result := -1;
   end;
+  if Result = 0 then
+    Result := AVERROR_EOF;
 end;
 
 {$IF LIBAVFORMAT_VERSION >= 54029104}

--- a/src/media/UMediaCore_FFmpeg.pas
+++ b/src/media/UMediaCore_FFmpeg.pas
@@ -602,17 +602,33 @@ begin
   if (Packet = nil) then
     Exit;
 
-  if (PChar(Packet^.data) <> STATUS_PACKET) then
-  begin
-    if (av_dup_packet(Packet) < 0) then
-      Exit;
-  end;
-
   CurrentListEntry := av_malloc(SizeOf(TAVPacketList));
   if (CurrentListEntry = nil) then
     Exit;
 
-  CurrentListEntry^.pkt  := Packet^;
+  {$IF LIBAVFORMAT_VERSION < 59000000}
+  CurrentListEntry^.pkt := Packet^;
+  if (PChar(Packet^.data) <> STATUS_PACKET) then
+  begin
+    if (av_dup_packet(@(CurrentListEntry^.pkt)) < 0) then
+    begin
+      av_free(CurrentListEntry);
+      Exit;
+    end;
+  end;
+  {$ELSE}
+  if (PChar(Packet^.data) <> STATUS_PACKET) then
+    begin
+      if (av_packet_ref(@CurrentListEntry^.pkt, Packet) < 0) then
+      begin
+        av_free(CurrentListEntry);
+        Exit;
+      end;
+    end
+  else
+    CurrentListEntry^.pkt := Packet^;
+  {$ENDIF}
+
   CurrentListEntry^.next := nil;
 
   SDL_LockMutex(Mutex);
@@ -753,7 +769,12 @@ begin
     if (PChar(CurrentListEntry^.pkt.data) = STATUS_PACKET) then
       FreeStatusInfo(CurrentListEntry^.pkt);
     // free packet data
+    {$IF LIBAVFORMAT_VERSION >= 59000000}
+    if (PChar(CurrentListEntry^.pkt.data) <> STATUS_PACKET) then
+      av_packet_unref(@CurrentListEntry^.pkt);
+    {$ELSE}
     av_free_packet(@CurrentListEntry^.pkt);
+    {$ENDIF}
     // Note: param must be a pointer to a pointer!
     av_freep(@CurrentListEntry);
     CurrentListEntry := TempListEntry;

--- a/src/media/UMediaCore_FFmpeg.pas
+++ b/src/media/UMediaCore_FFmpeg.pas
@@ -98,6 +98,7 @@ type
       procedure UnlockAVCodec();
       function AVFormatOpenInput(ps: PPAVFormatContext; filename: {const} PAnsiChar): Integer;
       procedure AVFormatCloseInput(ps: PPAVFormatContext);
+      function GetCodecContext(stream: PAVStream; codec: PAVCodec): PAVCodecContext;
   end;
 
 implementation
@@ -564,6 +565,19 @@ begin
   { avformat_close_input frees AVIOContext pb, no avio_close needed }
   { avformat_close_input frees AVFormatContext, no additional avformat_free_context needed }
   avformat_close_input(ps);
+  {$ENDIF}
+end;
+
+function TMediaCore_FFmpeg.GetCodecContext(stream: PAVStream; codec: PAVCodec): PAVCodecContext;
+begin
+  {$IF LIBAVFORMAT_VERSION < 59000000}
+  Result := stream^.codec;
+  {$ELSE}
+  Result := avcodec_alloc_context3(codec);
+  if (Result <> nil) and (avcodec_parameters_to_context(Result, stream^.codecpar) < 0) then
+    avcodec_free_context(@Result);
+  if Result = nil then
+    Log.LogError('Failed to allocate context', 'TMediaCore_FFmpeg.GetCodecContext');
   {$ENDIF}
 end;
 

--- a/src/media/UMediaCore_FFmpeg.pas
+++ b/src/media/UMediaCore_FFmpeg.pas
@@ -300,7 +300,7 @@ begin
       FirstAudioStream := i;
     end;
   end;
-{$ELSE}
+{$ELSEIF LIBAVFORMAT_VERSION < 59000000}
     if (Stream.codec.codec_type = AVMEDIA_TYPE_VIDEO) and
        (FirstVideoStream < 0) then
     begin
@@ -308,6 +308,19 @@ begin
     end;
 
     if (Stream.codec.codec_type = AVMEDIA_TYPE_AUDIO) and
+       (FirstAudioStream < 0) then
+    begin
+      FirstAudioStream := i;
+    end;
+  end;
+{$ELSE}
+    if (Stream.codecpar.codec_type = AVMEDIA_TYPE_VIDEO) and
+       (FirstVideoStream < 0) then
+    begin
+      FirstVideoStream := i;
+    end;
+
+    if (Stream.codecpar.codec_type = AVMEDIA_TYPE_AUDIO) and
        (FirstAudioStream < 0) then
     begin
       FirstAudioStream := i;
@@ -339,8 +352,10 @@ begin
 
 {$IF LIBAVCODEC_VERSION < 52064000} // < 52.64.0
     if (Stream.codec^.codec_type = CODEC_TYPE_AUDIO) then
-{$ELSE}
+{$ELSEIF LIBAVFORMAT_VERSION < 59000000}
     if (Stream.codec^.codec_type = AVMEDIA_TYPE_AUDIO) then
+{$ELSE}
+    if (Stream.codecpar^.codec_type = AVMEDIA_TYPE_AUDIO) then
 {$IFEND}
     begin
       StreamIndex := i;

--- a/src/media/UVideo.pas
+++ b/src/media/UVideo.pas
@@ -1149,8 +1149,10 @@ begin
   {$ELSEIF LIBAVCODEC_VERSION < 51105000}
   fCodecContext^.reordered_opaque := AV_NOPTS_VALUE;
   pts := fAVFrame^.reordered_opaque;
-  {$ELSE}
+  {$ELSEIF LIBAVCODEC_VERSION < 57061100}
   pts := fAVFrame^.pkt_pts;
+  {$ELSE}
+  pts := fAVFrame^.pts;
   {$IFEND}
 
   {$IF LIBAVCODEC_VERSION >= 51106000}

--- a/src/media/UVideo.pas
+++ b/src/media/UVideo.pas
@@ -571,9 +571,7 @@ var
   AudioStreamIndex: integer;
   r_frame_rate: cdouble;
   PossibleCodecs: array of PAVCodec;
-{$IF LIBAVCODEC_VERSION >= 58009100}
-  CodecIterator: pointer;
-{$IFEND}
+  CodecID: TAVCodecID;
 
   function ReduceSize: boolean;
   var
@@ -653,6 +651,12 @@ begin
 {$ELSE}
   fStream := PPAVStream(PtrUInt(fFormatContext^.streams) + fStreamIndex * Sizeof(pointer))^;
 {$IFEND}
+{$IF LIBAVFORMAT_VERSION < 59000000}
+  CodecID := fStream^.codec^.codec_id;
+{$ELSE}
+  CodecID := fStream^.codecpar^.codec_id;
+{$ENDIF}
+
   fCodecContext := fStream^.codec;
 
   fPreferDTS := false;
@@ -662,12 +666,12 @@ begin
 
   SetLength(PossibleCodecs, 1);
   for fCodec in PreferredCodecs do
-    if fCodec^.id = fCodecContext^.codec_id then
+    if fCodec^.id = CodecID then
     begin
       PossibleCodecs[High(PossibleCodecs)] := fCodec;
       SetLength(PossibleCodecs, Length(PossibleCodecs) + 1);
     end;
-  PossibleCodecs[High(PossibleCodecs)] := avcodec_find_decoder(fCodecContext^.codec_id);
+  PossibleCodecs[High(PossibleCodecs)] := avcodec_find_decoder(CodecID);
 
 {$IF LIBAVCODEC_VERSION >= 4008}
   fCodecContext^.get_format := @SelectFormat;

--- a/src/media/UVideo.pas
+++ b/src/media/UVideo.pas
@@ -1055,7 +1055,7 @@ begin
         fEOF := true;
         Exit;
       end;
-      if (errnum = -541478725) then // LIBAVFORMAT_VERSION >= 56000000 tells us if EOF reached.
+      if (errnum = AVERROR_EOF) then // LIBAVFORMAT_VERSION >= 56000000 tells us if EOF reached.
       begin
         fEOF := true;
         Exit

--- a/src/media/UVideo.pas
+++ b/src/media/UVideo.pas
@@ -637,7 +637,6 @@ begin
     Close();
     Exit;
   end;
-  Log.LogInfo('VideoStreamIndex : ' + inttostr(fStreamIndex), 'TVideoPlayback_ffmpeg.Open');
 
   // find video stream
   FFmpegCore.FindStreamIDs(fFormatContext, fStreamIndex, AudioStreamIndex);
@@ -647,6 +646,7 @@ begin
     Close();
     Exit;
   end;
+  Log.LogInfo('VideoStreamIndex : ' + inttostr(fStreamIndex), 'TVideoPlayback_ffmpeg.Open');
 
 {$IF LIBAVFORMAT_VERSION <= 52111000} // <= 52.111.0
   fStream := fFormatContext^.streams[fStreamIndex];

--- a/src/media/UVideo.pas
+++ b/src/media/UVideo.pas
@@ -657,8 +657,6 @@ begin
   CodecID := fStream^.codecpar^.codec_id;
 {$ENDIF}
 
-  fCodecContext := fStream^.codec;
-
   fPreferDTS := false;
   // workaround for FFmpeg bug #6560
   if (LeftStr(fFormatContext^.iformat^.name, 4) = 'mov,') or (fFormatContext^.iformat^.name = 'mov') then
@@ -673,29 +671,34 @@ begin
     end;
   PossibleCodecs[High(PossibleCodecs)] := avcodec_find_decoder(CodecID);
 
-{$IF LIBAVCODEC_VERSION >= 4008}
-  fCodecContext^.get_format := @SelectFormat;
-{$ENDIF}
-
-  // set debug options
-  fCodecContext^.debug_mv := 0;
-  fCodecContext^.debug := 0;
-
-  // detect bug-workarounds automatically
-  fCodecContext^.workaround_bugs := FF_BUG_AUTODETECT;
-  // error resilience strategy (careful/compliant/agressive/very_aggressive)
-  //fCodecContext^.error_resilience := FF_ER_CAREFUL; //FF_ER_COMPLIANT;
-  // allow non spec compliant speedup tricks.
-
-  //fCodecContext^.flags2 := CODEC_FLAG2_FAST;
-
-  // Note: avcodec_open() and avcodec_close() are not thread-safe and will
-  // fail if called concurrently by different threads.
   errnum := -1;
   for fCodec in PossibleCodecs do
   begin
     if fCodec = nil then
       continue;
+
+    fCodecContext := FFmpegCore.GetCodecContext(fStream, fCodec);
+    if fCodecContext = nil then
+      continue;
+
+    {$IF LIBAVCODEC_VERSION >= 4008}
+    fCodecContext^.get_format := @SelectFormat;
+    {$ENDIF}
+
+    // set debug options
+    fCodecContext^.debug_mv := 0;
+    fCodecContext^.debug := 0;
+
+    // detect bug-workarounds automatically
+    fCodecContext^.workaround_bugs := FF_BUG_AUTODETECT;
+    // error resilience strategy (careful/compliant/agressive/very_aggressive)
+    //fCodecContext^.error_resilience := FF_ER_CAREFUL; //FF_ER_COMPLIANT;
+    // allow non spec compliant speedup tricks.
+
+    //fCodecContext^.flags2 := CODEC_FLAG2_FAST;
+
+    // Note: avcodec_open() and avcodec_close() are not thread-safe and will
+    // fail if called concurrently by different threads.
     FFmpegCore.LockAVCodec();
     try
         {$IF LIBAVCODEC_VERSION >= 53005000}
@@ -709,6 +712,9 @@ begin
     if errnum >= 0 then
       Break;
     Log.LogError('Error ' + IntToStr(errnum) + ' returned by ' + fCodec^.name, 'TVideoPlayback_ffmpeg.Open');
+    {$IF LIBAVFORMAT_VERSION >= 59000000}
+    avcodec_free_context(@fCodecContext);
+    {$ENDIF}
   end;
   if (errnum < 0) then
   begin
@@ -947,7 +953,11 @@ begin
     // avcodec_close() is not thread-safe
     FFmpegCore.LockAVCodec();
     try
+      {$IF LIBAVFORMAT_VERSION < 59000000)}
       avcodec_close(fCodecContext);
+      {$ELSE}
+      avcodec_free_context(@fCodecContext);
+      {$ENDIF}
     finally
       FFmpegCore.UnlockAVCodec();
     end;
@@ -962,7 +972,9 @@ begin
     FFmpegCore.AVFormatCloseInput(@fFormatContext);
     {$IFEND}
 
+  {$IF LIBAVFORMAT_VERSION < 59000000)}
   fCodecContext  := nil;
+  {$ENDIF}
   fFormatContext := nil;
 
   if (fPboId <> 0) then

--- a/src/media/UVideo.pas
+++ b/src/media/UVideo.pas
@@ -801,6 +801,7 @@ begin
   fAVFrame := avcodec_alloc_frame();
   fAVFrameRGB := avcodec_alloc_frame();
   {$ENDIF}
+  {$IF LIBAVCODEC_VERSION < 59000000}
   fFrameBuffer := av_malloc(avpicture_get_size(PIXEL_FMT_FFMPEG,
       (fScaledWidth + BUFFER_ALIGN - 1) and -BUFFER_ALIGN, fScaledHeight));
 
@@ -819,6 +820,24 @@ begin
     Close();
     Exit;
   end;
+  {$ELSE}
+  if ((fAVFrame = nil) or (fAVFrameRGB = nil)) then
+  begin
+    Log.LogError('Failed to allocate buffers', 'TVideoPlayback_ffmpeg.Open');
+    Close();
+    Exit;
+  end;
+
+  errnum := av_image_alloc(@fAVFrameRGB^.data[0], @fAVFrameRGB^.linesize[0],
+                           (fScaledWidth + BUFFER_ALIGN - 1) and -BUFFER_ALIGN, fScaledHeight,
+                           PIXEL_FMT_FFMPEG, BUFFER_ALIGN);
+  if (errnum < 0) then
+  begin
+    Log.LogError('av_image_alloc failed: ' + FFmpegCore.GetErrorString(errnum), 'TVideoPlayback_ffmpeg.Open');
+    Close();
+    Exit;
+  end;
+  {$ENDIF}
 
   {$IFDEF UseSWScale}
   // if available get a SWScale-context -> faster than the deprecated img_convert().

--- a/src/media/UVideo.pas
+++ b/src/media/UVideo.pas
@@ -1059,8 +1059,13 @@ begin
       if (avio_feof(pbIOCtx) <> 0) then
       {$IFEND}
       begin
+        {$IF LIBAVCODEC_VERSION < 57037100}
         fEOF := true;
         Exit;
+        {$ELSE}
+        avcodec_send_packet(fCodecContext, nil);
+        Continue;
+        {$ENDIF}
       end;
 
       // check for errors
@@ -1082,15 +1087,16 @@ begin
       {$ELSE}
       fileSize := fFormatContext^.file_size;
       {$IFEND}
-      if ((fileSize <> 0) and (pbIOCtx^.pos >= fileSize)) then
+      if (((fileSize <> 0) and (pbIOCtx^.pos >= fileSize))
+          or (errnum = AVERROR_EOF)) then // LIBAVFORMAT_VERSION >= 56000000 tells us if EOF reached.
       begin
+        {$IF LIBAVCODEC_VERSION < 57037100}
         fEOF := true;
         Exit;
-      end;
-      if (errnum = AVERROR_EOF) then // LIBAVFORMAT_VERSION >= 56000000 tells us if EOF reached.
-      begin
-        fEOF := true;
-        Exit
+        {$ELSE}
+        avcodec_send_packet(fCodecContext, nil);
+        Continue;
+        {$ENDIF}
       end;
 
       // error occured, log and exit

--- a/src/media/UVideo.pas
+++ b/src/media/UVideo.pas
@@ -1028,6 +1028,19 @@ begin
   // read packets until we have a finished frame (or there are no more packets)
   while (FrameFinished = 0) do
   begin
+    {$IF LIBAVCODEC_VERSION >= 57037100}
+    errnum := avcodec_receive_frame(fCodecContext, fAVFrame);
+    if errnum = 0 then
+      Break;
+    if errnum = AVERROR_EOF then
+    begin
+      fEOF := true;
+      Exit;
+    end;
+    if errnum <> AVERROR(EAGAIN) then
+      Continue;
+    {$ENDIF}
+
     errnum := av_read_frame(fFormatContext, AVPacket);
     if (errnum < 0) then
     begin
@@ -1100,9 +1113,11 @@ begin
       {$IF LIBAVFORMAT_VERSION < 52012200)}
       avcodec_decode_video(fCodecContext, fAVFrame,
           frameFinished, AVPacket.data, AVPacket.size);
-      {$ELSE}
+      {$ELSEIF LIBAVCODEC_VERSION < 57037100}
       avcodec_decode_video2(fCodecContext, fAVFrame,
           frameFinished, @AVPacket);
+      {$ELSE}
+      avcodec_send_packet(fCodecContext, @AVPacket);
       {$IFEND}
 
       {$IF LIBAVCODEC_VERSION < 51106000}

--- a/src/media/UVideo.pas
+++ b/src/media/UVideo.pas
@@ -526,7 +526,9 @@ begin
 
   FFmpegCore := TMediaCore_FFmpeg.GetInstance();
 
+  {$IF LIBAVFORMAT_VERSION < 58027100}
   av_register_all();
+  {$ENDIF}
 end;
 
 function TVideoPlayback_FFmpeg.Finalize(): boolean;

--- a/src/media/UVideo.pas
+++ b/src/media/UVideo.pas
@@ -741,7 +741,7 @@ begin
     fAspect := fAspect * fCodecContext^.width /
                          fCodecContext^.height;
 
-  {$IF LIBAVFORMAT_VERSION_MAJOR >= 55}
+  {$IF (LIBAVFORMAT_VERSION_MAJOR >= 55) and (LIBAVFORMAT_VERSION_MAJOR < 59)}
   r_frame_rate := av_q2d(av_stream_get_r_frame_rate(fStream));
   {$ELSE}
   r_frame_rate := av_q2d(fStream^.r_frame_rate);

--- a/src/media/UVideo.pas
+++ b/src/media/UVideo.pas
@@ -1092,7 +1092,11 @@ begin
     end;
 
     // free the packet from av_read_frame
+    {$IF LIBAVCODEC_VERSION < 59000000}
     av_free_packet( @AVPacket );
+    {$ELSE}
+    av_packet_unref(@AVPacket);
+    {$ENDIF}
   end;
 
   // reset opaque data and update pts

